src/depends/libRumorSpreading/RumorStateMachine.cpp

```diff
@@ -0,0 +1,166 @@
+#include "RumorStateMachine.h"
+
+#define LITERAL(s) #s
+
+namespace RRS {
+
+// STATIC MEMBERS
+std::map<RumorStateMachine::State, std::string> RumorStateMachine::s_enumKeyToString = {
+    {UNKNOWN,    LITERAL(UNKNOWN)},
+    {NEW,        LITERAL(NEW)},
+    {KNOWN,      LITERAL(KNOWN)},
+    {OLD,        LITERAL(OLD)},
+    {NUM_STATES, LITERAL(NUM_STATES)}
+};
+
+
+// PRIVATE METHODS
+void RumorStateMachine::advanceNew(const std::unordered_set<int>& membersInRound)
+{
+    m_roundsInB++;
+    if (m_age >= m_networkConfigPtr->maxRoundsTotal()) {
+        advanceOld();
+        return;
+    }
+
+    for (auto id : membersInRound) {
+        if (m_memberRounds.count(id) <= 0) m_memberRounds[id] = 0;
+    }
+
+    // Compare our age to the majority of rounds
+    int numLess = 0;
+    int numGreaterOrEqual = 0;
+    for (const auto entry : m_memberRounds) {
+        int theirRound = entry.second;
+        if (theirRound < m_age) {
+            numLess++;
+        } else if (theirRound >= m_networkConfigPtr->maxRoundsInB()) {
+            m_state = KNOWN;
+        } else {
+            numGreaterOrEqual++;
+        }
+    }
+
+    if (numGreaterOrEqual > numLess) {
+        m_roundsInB++;
+    }
+
+    if (m_roundsInB >= m_networkConfigPtr->maxRoundsInB()) {
+        m_state = KNOWN;
+    }
+    m_memberRounds.clear();
+}
+
+void RumorStateMachine::advanceKnown()
+{
+    m_roundsInC++;
+    if (m_age >= m_networkConfigPtr->maxRoundsTotal() ||
+        m_roundsInC >= m_networkConfigPtr->maxRoundsInC()) {
+        advanceOld();
+    }
+
+}
+
+void RumorStateMachine::advanceOld()
```

Comment:
Function name vs context is a bit inconsistent with the two other functions:
advanceNew -> advance **from** NEW state
advanceKnown -> advance **from** KNOWN state
advanceOld -> advance **to** OLD state